### PR TITLE
Softmax

### LINKED ISSUE #369: [Relay] Toward a working VGG Net (continued).
## VGG Net
- [x] Change go-done interface to `invoke`ables. This requires [Dahlia #340](https://github.com/cucapra/dahlia/pull/340). (#379)
- [X] Re-add the rest of the supported Relay functions. (#396)
- [X] Support constants (#406)
- [x] Support `std_exp` round 2; finish `nn.softmax` implementation. This should be the last piece of lowering the entire VGG Net to FuTIL. (#396, #404, #490) 

## Code Improvements
- [x] Fix the test suite. Tests for individual Relay function calls (e.g. `nn.relu`) should be part of the integration test suite. However, a few unit tests should be added to ensure that any changes to Calyx affecting the Relay frontend will cause notice. (#396)
- [x] Fix FuTIL AST. Most of the work here should be done in #368.
- [X] Clean up `frontends/relay/example.py`. (#406)
- [x] Update installation. Currently missing `antlr-4-python3-runtime` pip install. Move to main Calyx documentation. (#408)

### PR DESCRIPTION
**Pull request by cgyurgyik**
The final piece to the `vgg_net` and `mlp_net` puzzle!!

```
import tensorflow as tf
import numpy as np

tf.nn.softmax(np.array([[1.0, 2.0, 3.0, 4.0], [8.0, 2.0, 4.0, 2.0]]), axis=1)
<tf.Tensor: shape=(2, 4), dtype=float64, numpy=
array([
  [
   0.0320586, 
   0.08714432, 
   0.23688282, 
   0.64391426
  ],
  [
   0.97725618, 
   0.00242238, 
   0.01789907, 
   0.00242238
  ]
])>
```

Compared to Calyx:
```
  "x1": [
    [
      "0.0320587158203125",
      "0.0871429443359375",
      "0.23687744140625",
      "0.6439056396484375"
    ],
    [
      "0.9772491455078125",
      "0.002410888671875",
      "0.0178985595703125",
      "0.002410888671875"
    ]
```
Next step is perhaps walking through a smaller net such as `mlp_net` component by component and comparing it to TF, as I'm sure bugs exist.

Closes #369.

### CODE REVIEW

**Comment by cgyurgyik** [comment]
Blocked on [Dahlia #368](https://github.com/cucapra/dahlia/pull/368).

File: examples/dahlia/dot-product.expect [hunk 1]
```diff
@@ -1,5 +1,5 @@
 {
-  "cycles": 91,
+  "cycles": 131,
```
Comment (rachitnigam) [inline examples/dahlia/dot-product.expect:2]: What caused this change? It looks like the Dahlia backend stopped generating `"static"` annotations for a bunch of groups it did before.

File: primitives/binary_operators.sv [hunk 2]
```diff
@@ -96,6 +96,17 @@ module std_fp_div_pipe #(
     end
 
     always_ff @(posedge clk) begin
+      if (!go) begin
```
Comment (rachitnigam) [inline primitives/binary_operators.sv:99]: Why is this needed?

File: examples/dahlia/dot-product.expect (hunk 1 above)
Comment (cgyurgyik) [reply examples/dahlia/dot-product.expect:2]: For [Dahlia 368](https://github.com/cucapra/dahlia/pull/368), in the static timing map, I removed the associations `mult -> 3 cycles` and `sqrt -> 17 cycles`. 

Perhaps I should've discussed this more beforehand, but my thought process below:

For `sqrt`, the number of iterations is now dependent on the WIDTH provided.
For `mult`, I wasn't able to find anything that says the multipliers for fixed point run in exactly 3 cycles.

File: primitives/binary_operators.sv (hunk 2 above)
Comment (cgyurgyik) [reply primitives/binary_operators.sv:99]: Without this, the expected output is incorrect; it just repeats the output of the first division. My initial guess was because a signal wasn't being reset, and alas this worked.

I'm happy to dig deeper into this with a waveform viewer, but I looked at your implementation of `std_div_pipe` and it follows a similar approach.

Comment (rachitnigam) [reply primitives/binary_operators.sv:99]: No, I think that's fine.

File: examples/dahlia/dot-product.expect (hunk 1 above)
Comment (rachitnigam) [reply examples/dahlia/dot-product.expect:2]: The multiplier implementation in the standard library takes three cycles. This is important for targeting FPGAs since multipliers are mapped on DSP blocks which must take at least three cycles to perform the multiplication.

If that makes sense, I'd recommend adding back the `mult` stuff. The rationale for `sqrt` is reasonable so no need to change it.

Comment (cgyurgyik) [reply examples/dahlia/dot-product.expect:2]: Done in [Dahlia 370](https://github.com/cucapra/dahlia/pull/370).

**Comment by rachitnigam** [comment]
LGTM!
